framework/src/geomsearch/NearestNodeLocator.C

```diff
@@ -69,8 +70,13 @@ NearestNodeLocator::findNodes()
    */
   const std::map<dof_id_type, std::vector<dof_id_type>> & node_to_elem_map = _mesh.nodeToElemMap();
 
-  if (_first)
+  if (_first || (_reinit_iteration && _patch_update_strategy == Moose::Iteration))
   {
+    // After a call from system reinit, mesh has been updated with initial adaptivity.
+    // Moose::Iteration relies on data generated for ghosting (i.e. trial_primary_nodes)
+    if (!_first)
+      _reinit_iteration = false;
```

Comment:
I m not sure about this logic. 
Seems like we:
- findNodes on first
- the second time, we do it again because reinit_iteration started true and is still true
- the third time we dont because reinit_iteration is not true

should we just keep track of the number of iterations instead of these two booleans?

Reply:
The purpose is to correctly identify nearest nodes when we do initial adaptivity. To that effect, we need to go through that piece of code upon `reinit()`, in addition to `init()`. 

There may be other ways to address that problem. We could keep track of the number of iterations. However, that code gets called a few times before we attempt to the first nonlinear iteration. If we based the logic on the nonlinear iterations, we would probably be going through that code more often than needed.

All that said, I am not very familiar with how initial adaptivity or adaptivity in general is supposed to work with the calls within `FEProblemBase`, so I'd be happy to change the logic.

Reply:
I dont see any logic that captures this:
```
The purpose is to correctly identify nearest nodes when we do initial adaptivity. To that effect, we need to go through that piece of code upon reinit(), in addition to init().
```

Reply:
This code gets called a second time on reinit. That's enabled by the reinit flag you commented on above.

Reply:
can you set this boolean on reinit() instead?